### PR TITLE
PAT-325: Full needs priority evaluation feature

### PR DESCRIPTION
**Pull request by timharrison-moj**
This adds a new table, new routes and forms, new needs banding algorithm, and new question configuration file (similar to risk-banding).
This is a first-cut of the needs priority evaluation as there are some questions outstanding and work which I know needs to be added in : 
* Do we need to prevent access if a risk-banding has not been completed?
* What should happen to the needs priority scores if a risk banding is subsequently changed?
* We do not calculate the TAPS and BPS score yet as these rely on a completed risk banding
* I need to add gender into the basic offender details - one of the needs scores requires it - defaulted to Male at present.
* The user-guide wording needs some content input on the mental health and behaviours pages - just placeholder text for now.
* Some of the form and route functions could be shared between needs and risk-banding - this is an additional task to take on once this PR is in.

### CODE REVIEW

File: server/services/pathfinderService.js [hunk 1]
```diff
@@ -185,6 +185,37 @@ module.exports = function createPathfinderService({ pathfinderClient, elite2Clie
     return auditId
   }
 
+  async function createNeedsPriorities(nominalId, form, username) {
+    const needsPrioritiesId = await pathfinderClient.createNeedsPriorities(nominalId, form, username)
+    return needsPrioritiesId
+  }
+
+  async function getLatestNeedsPriorities(nominalId, username) {
```
Comment (yasinmustafa) [inline server/services/pathfinderService.js]: the method name here is a little deceptive.. the actual code may create a Needs Priority.. either rename the method name or add some method comments to explain

File: server/views/formPages/needs/needsForm.html [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+{% from "govuk/components/back-link/macro.njk" import govukBackLink %}
+{% from "govuk/components/checkboxes/macro.njk" import govukCheckboxes %}
+{% from "govuk/components/radios/macro.njk" import govukRadios %}
+{% from "govuk/components/input/macro.njk" import govukInput %}
+
+{% block content %}
+
+{{ govukBackLink({
+  text: "Back",
+  href: "/taskList/" + data.nominalId
```
Comment (yasinmustafa) [inline server/views/formPages/needs/needsForm.html]: this shouldn't work as there isn't a taskList any more!
should be 
`href: "/taskList/" + data.nominalId`

File: server/views/pages/needs/acknowledge.html [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+          classes: "govuk-button--primary",
+          text: "Continue to controls and interventions",
+          href: "/nominal/cim-matrix/" + nominalId,
+          preventDoubleClick: true
+          })
+        }}
+
+        {{ govukButton({
+          classes: "govuk-button--secondary",
+          text: "Return to task list",
```
Comment (yasinmustafa) [inline server/views/pages/needs/acknowledge.html]: Should the text really be 'Return to  task list'>

File: server/views/pages/needs/acknowledge.html [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+          href: "/nominal/cim-matrix/" + nominalId,
```
Comment (yasinmustafa) [inline server/views/pages/needs/acknowledge.html]: The href here doesn't seem right... it's the same as the other button

File: migrations/20191203120000_create_needs_priorities.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+exports.up = knex =>
+  Promise.all([
+    knex.schema.createTable('needs_priorities', table => {
+      table.increments('id').primary('pk_needs_priorities')
+      table.dateTime('started_time')
+      table.dateTime('completed_time')
+      table.string('needs_status')
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply migrations/20191203120000_create_needs_priorities.js]: A bit wary of using "status" as a column name as it's often a language or db reserved word. I did this intentionally to keep it from being called status.

File: server/config/needs/needsBanding.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
+const R = require('ramda')
+const logger = require('../../../log.js')
+const { isNilOrEmpty } = require('../../utils/utils')
+
+const getRecommendedNeedsBand = (needsFormResponses, gender) => {
+  let priority = 'NA'
+  let score = 0
+
+  try {
+    const needsAttributesPresent = R.flatten(getNeedsAttributesFromForm(needsFormResponses))
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/config/needs/needsBanding.js]: Thanks!

File: server/data/pathfinderClient.js [hunk 7]
```diff
@@ -342,6 +342,75 @@ const getHighestPreviousBand = async nominalId => {
   return result.rows[0].highestBand
 }
 
+const createNeedsPriorities = async (nominalId, needsForm, username) => {
+  const result = await nonTransactionalClient.query({
+    text: `insert into needs_priorities 
+           (nominal_id, needs_status, needs_form, started_time, last_updated_by)
+           values ($1, 'CREATED', $2, now(), $3)
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/data/pathfinderClient.js]: Yes, I will look to do this tonight.

File: server/data/pathfinderClient.js [hunk 8]
```diff
@@ -342,6 +342,75 @@ const getHighestPreviousBand = async nominalId => {
...
+    values: [nominalId],
+  })
+  return result.rows[0]
+}
+
+const updateNeedsPriorities = async (nominalId, needsForm, username) => {
+  const result = await nonTransactionalClient.query({
+    text: `update needs_priorities np1
+           set needs_form = $2,
+               needs_status = 'IN PROGRESS',
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/data/pathfinderClient.js]: Ditto.

File: server/views/formPages/needs/needsForm.html (hunk 2 above)
Comment (timharrison-moj) [reply server/views/formPages/needs/needsForm.html]: Haha. Yes, I actually noticed this when I tested and made a note to change it, then forgot. Changed now - not pushed yet,

File: server/views/pages/needs/acknowledge.html (hunk 3 above)
Comment (timharrison-moj) [reply server/views/pages/needs/acknowledge.html]: Ahem, No. Will update that.

File: server/views/pages/needs/acknowledge.html (hunk 4 above)
Comment (timharrison-moj) [reply server/views/pages/needs/acknowledge.html]: Will check and correct now.

File: migrations/20191203120000_create_needs_priorities.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      table.string('priority')
+      table.integer('score')
+      table.integer('bps_score')
+      table.integer('taps_score')
+      table.string('config_version')
+      table.jsonb('needs_form')
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply migrations/20191203120000_create_needs_priorities.js]: No danger of reserved word or method here, so will update it to form.

File: server/views/pages/needs/acknowledge.html (hunk 3 above)
Comment (timharrison-moj) [reply server/views/pages/needs/acknowledge.html]: Now done.

File: server/views/pages/needs/acknowledge.html (hunk 4 above)
Comment (timharrison-moj) [reply server/views/pages/needs/acknowledge.html]: Ah - We don't have any interventions feature yet - so just points back at the same CIM page.

File: server/services/pathfinderService.js (hunk 1 above)
Comment (timharrison-moj) [reply server/services/pathfinderService.js]: Renamed to getOrCreateLatestNeedsPriorities.

File: server/services/pathfinderService.js [hunk 10]
```diff
@@ -2,7 +2,8 @@ const R = require('ramda')
 const { isNilOrEmpty, properCaseName, formatDate } = require('../utils/utils')
 const riskBanding = require('../config/banding/riskBanding')
 const riskBandingStatus = require('../config/riskBandingStatus')
-const logger = require('../../log.js')
+const needsBanding = require('../config/needs/needsBanding')
+// const logger = require('../../log')
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/services/pathfinderService.js]: Done.

File: migrations/20191203120000_create_needs_priorities.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply migrations/20191203120000_create_needs_priorities.js]: Done this one.

[the remark this answers is not in the document]
Comment (timharrison-moj) [reply migrations/20191203120000_create_needs_priorities.js]: Done this one.

File: migrations/20191203120000_create_needs_priorities.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply migrations/20191203120000_create_needs_priorities.js]: I changed this as it does not fall foul any any reserved words here.

File: server/routes/createNeeds.js [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+  /**
+   * This function is used recursively to determine if all tasks are complete or if a single task is complete (including sub-tasks).
+   * It uses the answers in the form and compares to the config.
+   *
+   * @param sectionId
+   * @param formAnswers
+   * @param onlyCheckOneSection - if true then only the section passed in will be checked (including sub-tasks)
+   * @returns {*|boolean}
+   */
+  // TODO: Make common function to share with risk banding and other config-based questionaires - formProcessing?
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/createNeeds.js]: Yeah, I have a list of follow-on work and questions.

File: server/routes/needsIndex.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
+const asyncMiddleware = require('../middleware/asyncMiddleware')
+const PathfinderNeedsRoutes = require('./createNeeds')
+const needsConfig = require('../config/needs/v1')
+// const logger = require('../../log')
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/needsIndex.js]: Done.

File: server/routes/createNeeds.js [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+      }
+
+      // Get the priority and score, and finalise the needs priorities if all tasks are complete
+      if (allNeedsTasksComplete) {
+        // TODO: Get the gender added to the offenderBasic details so that it is available here, defaults to 'M'
+        logger.info(`Offender basic needs to contain a gender value for determineNeedsBand() - ${offenderBasic.nomsId}`)
+        const needs = await pathfinderService.calculateNeedsBand(nominalId, mergedPayload, 'M')
+        logger.info(`Finalising needs priorities with priority: ${needs.priority}, score: ${needs.score}`)
+
+        // TODO: Bolster error handling/exceptions around service calls
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/createNeeds.js]: I wondered about catching exceptions here and acting on them, but have a note to go back and check various scenarios here.

File: server/data/pathfinderClient.js [hunk 14]
```diff
@@ -342,6 +342,75 @@ const getHighestPreviousBand = async nominalId => {
...
+  return result
+}
+
+const finaliseNeedsPriorities = async (nominalId, needsForm, priority, score, username) => {
+  const result = await nonTransactionalClient.query({
+    text: `update needs_priorities np1
+           set needs_form = $2,
+               priority = $3,
+               score = $4,
+               needs_status = 'COMPLETED',
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/data/pathfinderClient.js]: done

File: server/data/pathfinderClient.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/data/pathfinderClient.js]: done

**Pull request merged**